### PR TITLE
Introduce streaming for ConvertFrom-Json

### PR DESCRIPTION
**Pull request by strawgate**
# PR Summary

This PR performs special handling for a particular json use-case where the response body is an array. This detects that situation and deserializes the array one object at a time. This greatly improves runtime characteristics and reduces memory usage when the json is an array of objects while preventing us from having to implement our own full json parser.

This PR does not improve performance of parsing a single large dictionary.

## PR Context

ConvertFrom-Json is a core cmdlet for anyone using Powershell for Web APIs. The Cmdlet has very poor runtime characteristics. It takes a long time to deserialize json compared to other deserializing in other languages and it uses a lot of memory to deserialize (15x the size of the json).

This PR on its own increases memory usage as compared to master. Combined with https://github.com/PowerShell/PowerShell/pull/15684 this PR cuts memory usage and runtime by more than half.

Benchmark for 100MB JSON file:

Master
- Memory Usage: **1079.3671875 MB**
- Runtime: **7256.4268 ms**

ThisPR Alone
- Memory Usage: 1477.83203125 MB
- Runtime: 6574.552 ms

https://github.com/PowerShell/PowerShell/pull/15684
- Memory Usage: 1077.58203125 MB
- Runtime: 3885.0489 ms

ThisPR + https://github.com/PowerShell/PowerShell/pull/15684
- Memory Usage: **451.390625 MB**
- Runtime: **3077.3673 ms**

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs [hunk 1]
```diff
@@ -160,46 +161,69 @@ public static object ConvertFromJson(string input, bool returnHashtable, int? ma
...
+                using (System.IO.MemoryStream stream = new System.IO.MemoryStream(System.Text.Encoding.Default.GetBytes(input)))
+                {
+                    using (System.IO.StreamReader streamReader = new System.IO.StreamReader(stream))
                     {
-                        // This TypeNameHandling setting is required to be secure.
-                        TypeNameHandling = TypeNameHandling.None,
-                        MetadataPropertyHandling = MetadataPropertyHandling.Ignore,
-                        MaxDepth = maxDepth
-                    });
+                        using (JsonReader reader = new JsonTextReader(streamReader))
```
Comment (Emdot) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs:174]: I'm not sure which version of C# this project uses, but I _think_ it's C# 9, in which case you can use
```
using System.IO.MemoryStream stream = new System.IO.MemoryStream(System.Text.Encoding.Default.GetBytes(input));
using System.IO.StreamReader streamReader = new System.IO.StreamReader(stream);
using JsonReader reader = new JsonTextReader(streamReader);
```
instead to reduce a lot of the nesting.

Comment (strawgate) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs:174]: done!

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs [hunk 2]
```diff
@@ -160,46 +161,69 @@ public static object ConvertFromJson(string input, bool returnHashtable, int? ma
...
+                        {
+                            bool isArray = false;
+                            var readResult = reader.Read();
+                            // If the first token in our file is an array let's read in that token so that our next token is an object or a primitive.
+                            // This will allow newtonsoft to deserialize the incoming json one object at a time instead of doing the whole object at once.
+                            if (reader.TokenType == JsonToken.StartArray)
```
Comment (Emdot) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs]: Could this be something like (pseudocode):

```
if token type is startarray
    return ParseArray(serializer, reader, out error)
else
    return ParseStandard(serialize, reader, out error)

ParseStandard:
    call serializer.Deserialize
    switch obj
        each case returns a value, as in the original code

ParseArray
    elements = new List<object>()
    while reader.Read and token isn't end-array
        originalElement = serializer.Deserialize
        processedElement = ParseStandard(serializer, reader, out error)
        if error break
    return elements.ToArray
```

Comment (strawgate) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs]: My latest commit should refactor it to be more like this

**Pull request closed without merging**